Adjusted implementation of sqlite3_rekey_v2 to use a modified version of sqlite3RunVacuum

utelle opened the pull request:
This allows to regain reserved page space on decrypting an encrypted database (solving issue #9).

Additionally issue #10 is fixed.

Optionally keep bytes 16 to 23 of the database header unencrypted (symbol BYTE16TO23_UNENCRYPTED), if compatibility with official SQLite is required (see issue #8).

resilar commented:
Thanks a lot. The code seems to be working fine, including the regaining of reserved space that I thought would corrupt DBs. There are some minor style inconsistencies in this commit and also in existing sqleet code, which I'll address in a future commit. Mainly:

* Try to stay within 80 columns
* Commit message could be [prettier](https://chris.beams.io/posts/git-commit/)

Do not close the issues #8 #9 #10 yet. I can do that after reading through them and writing my comments (if I have any) later tonight.

utelle commented:
> Thanks a lot.

You are welcome. Since I'm currently working on extending my own SQLite encryption extension (especially incorporating the sqleet encryption scheme), it was no big deal to prepare this patch for sqleet.

> The code seems to be working fine, including the regaining of reserved space that I thought would corrupt DBs.

Up to now I didn't experience database corruption. However, I don't know the SQLite internals good enough to understand why the original SQLite `vacuum` code never ever reduces the number of reserved bytes. So there might exist situations where this can lead to problems.

> There are some minor style inconsistencies in this commit and also in existing sqleet code, which I'll address in a future commit.

I tried to follow the style I found in sqleet. However, for example I just added a few comments to the modified code of function `sqlite3RunVacuum` to describe what I changed, but left the rest of the code untouched (as copied from SQLite).

> - Try to stay within 80 columns

Well ... the curse of wide screens ...

> - Commit message could be prettier

I have to admit that I wasn't happy with my commit message myself either. I'll try to improve in the future.

> Do not close the issues #8 #9 #10 yet. I can do that after reading through them and writing my comments (if I have any) later tonight.

Ok. However, comments can always be added, even to a closed issue, unless the conversation is locked.
